contrib/win32/win32compat/shell-host.c

```diff
@@ -231,26 +263,80 @@ initialize_keylen()
 		keys[i].in_key_len = (int) wcslen(keys[i].in);
 }
 
+int
+ProcessCtrlSequence(wchar_t *buf, int buf_len)
+{
+	int vkey = 0;
+	/* Decode special keys when pressed CTRL key */
+	if (buf[0] == L'\033' && buf[1] == L'[' && buf[buf_len - 3] == L';' && buf[buf_len - 2] == L'5') {
+		if (buf[buf_len - 1] == L'~') {
+			/* VK_DELETE, VK_PGDN, VK_PGUP */
+			if (!vkey && buf_len == 6)
+				vkey = GetVirtualKeyByMask(L'[', &buf[2], 1, L'~');
+
+			/* VK_F1 ... VK_F12 */
+			if (!vkey && buf_len == 7)
+				vkey = GetVirtualKeyByMask(L'[', &buf[2], 2, L'~');
+		} else {
+			/* VK_LEFT, VK_RIGHT, VK_UP, VK_DOWN */
+			if (!vkey && buf_len == 6 && buf[2] == L'1')
+				vkey = GetVirtualKeyByMask(L'[', &buf[5], 1, 0);
+
+			/* VK_F1 ... VK_F4 */
+			if (!vkey && buf_len == 6 && buf[2] == L'1' && isalpha(buf[5]))
+				vkey = GetVirtualKeyByMask(L'O', &buf[5], 1, 0);
+		}
+		if (vkey) {
+			SendKeyStrokeEx(child_in, VK_CONTROL, 0, LEFT_CTRL_PRESSED, TRUE);
+			SendKeyStrokeEx(child_in, vkey, 0, LEFT_CTRL_PRESSED, TRUE);
+			SendKeyStrokeEx(child_in, VK_CONTROL, 0, 0, FALSE);
+			SendKeyStrokeEx(child_in, vkey, 0, 0, FALSE);			
+		}
+	}
+
+	return vkey;
+}
+
 void 
 ProcessIncomingKeys(char * ansikey)
 {
 	wchar_t *buf = utf8_to_utf16(ansikey);
-		
+	int buf_len = 0;
```

Comment:
can you declare those variables before calling utf8_to_utf16, so the check on return can be right after the call?